Fix missing originOp on lifecycle metadata updates

francoisferrand opened the pull request:
Incorrect or missing `originOp` causes incorrectly duplicated bucket notifications, missing notifications or wrong notifications being sent.

- Fix bucket notification at the end of archive
- Clear originOp when updating replication status
- Fix flaky BackbeatConsumer rebalance test

Issue: BB-472

francoisferrand committed:
Fix bucket notification at the end of archive

Properly set originOp during MD operations for archive, to ensure the
proper notification (`s3:LifecycleTransition`) is sent.

Non-standard originOp's are set when starting the transition, retrying
or setting the archive field, but this will not generate any bucket
notification since the only S3 notification is `s3:LifecycleTransition`,
with no wildcard.

Issue: BB-472

francoisferrand committed:
Clear originOp when updating replication status

This ensures no extra bucket notification is sent.

Issue: BB-472

francoisferrand committed:
Fix flaky BackbeatConsumer rebalance test

Issue: BB-447

KillianG commented on extensions/replication/tasks/UpdateReplicationStatus.js:
```diff
@@ -246,6 +246,7 @@ class UpdateReplicationStatus extends BackbeatTask {
         }
         const versionId =
             sourceEntry.getReplicationSiteDataStoreVersionId(site);
+        entry.setOriginOp('');
```
For my understanding can you explain why we set an empty origin op ?

jonathan-gramain replied:
This is to avoid double notifications essentially, as keeping the original field value would trigger a new notification on update, but maybe we could/should set an actual value here instead like `s3:ReplicateObject` (if that's your original question).

KillianG replied:
`s3:ReplicateObject` is not [AWS standard](https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-event-types-and-destinations.html#:~:text=s3%3AReplication%3A*,s3%3AReplication%3AOperationNotTracked).. 
here are the only standard notifications for replication: 
```
s3:Replication:*
s3:Replication:OperationFailedReplication
s3:Replication:OperationMissedThreshold
s3:Replication:OperationReplicatedAfterThreshold
s3:Replication:OperationNotTracked
```

I don't see any that can fit our need here..

Can we introduce a non standard notification for that case (should we ?) ?

francoisferrand replied:
This is not a standard notification indeed, but will not generate any notification anyway : since it will not be possible to register for this kind of notification.

The purpose of adding an `originOp` in that case would be to make it easier to see (in oplog or while troubleshooting) the "reason" for the metadata write operation; and it makes it easier to handle the case where we want to introduce a notification in that case (we can simply allow registering for such notification, or easily identification this case and set the proper id).

We could thus set `s3:ReplicateObject`, which would solve the issue just as well and not create any divergence form standard, since does not start with `s3:Repication:` and will thus not be matched by `s3:Replication:*`.

jonathan-gramain reviewed (commented):
Left a comment replying to @KillianG , otherwise LGTM

francoisferrand committed:
Refresh metadata when setting transition in progress flag

This is needed to minimize the risk of race condion on metadata update
(see [1]). We used to read the object metadata, send a kafka message and
wait for delivery report (~2 seconds) before writing the updated
metadata: which leaves a significant window for the race condition.

[1] https://scality.atlassian.net/browse/ARSN-341

Issue: BB-472

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-472 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.31`

* `8.7.0`


Please check the `Fix Version/s` of BB-472, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

francoisferrand commented:
/approve

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-472` into
`development/8.6`
* [w/8.7/bugfix/BB-472](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/bugfix/BB-472) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



Please check the status of the associated issue BB-472.

Goodbye francoisferrand.
